lib/base-compiler.js

```diff
@@ -699,8 +693,16 @@ export class BaseCompiler {
         return inputFilename.replace(path.extname(inputFilename), '.ll');
     }
 
-    getRustMirOutputFilename(inputFilename) {
-        return inputFilename.replace(path.extname(inputFilename), '.mir');
+    getRustMirOutputFilename(rustcOptions, outputFilename) {
+        let emits = nopt({emit: Array}, {}, rustcOptions, 0).emit;
+        let non_mir_emits = _.filter(emits, emit => emit !== 'mir').length > 0;
+        if (non_mir_emits) {
+            let basename = path.basename(outputFilename).split('.')[0] + '.mir';
+            let dirname = path.dirname(outputFilename);
+            return path.join(dirname, basename);
+        } else {
+            return outputFilename;
+        }
```

Comment:
I'm not sure to really understand what is changing here. Could you maybe just add a comment to make it clearer ?

Reply:
Yeah, a comment will definitely help here. Bottom line, if you pass multiple `--emit` and also a `-o` then `rustc` will infer names for each output from the explicit names. The multiple `--emit` comes from prepending the rest of the flags.

Reply:
What is confusing me here is that the original code relies on `inputFilename` so that `--emit mir -o frodo.o boro.rs` writes in `boro.mir`... and now, it only use the output? Also I'm not quite sure (haven't read the full code yet), but is there a possible race here ? The original `rustc` execution will be executed in // with the execution for the MIR output. If both execution are sharing option, maybe both will try to write the same files ? Or is not the case ?

Reply:
Yes, the second call to `rustc` writes the same file again, but because the call has the same flags the file will have the same content. This is the same behavior as `generateIR` which just appends `--emit llvm-ir` and calls `rustc` again. In this case, I'm appending `--emit mir` instead. 

The only difference is that `getRustMimrOutputFilename` is being smarter about inferring the name of the generated `mir` file which depends on the combination of flags you're passing. The old code didn't have this problem because it had control over all the flags. `getIrOutputFilename` though is arguably wrong because it will fail if `--emit llvm-ir` is passed explicitly (https://godbolt.org/z/soeebjdcW). I think it will also fail if `backendOptions.customOutputFilename` is set to something with more than one dot in the name.

Reply:
Ok, thank you for the details :). I've been trying to test your code. It works fine, thanks. But I could not find a case that is currently broken in current code and is fixed in yours wrt to this output file. I don't doubt you are correct, I'm simply trying to see how it works :). Thanks for pointing out LLVM IR is broken, I'll open an issue

Reply:
After thinking a bit more about it, I think I understand now, so I'll test a bit more and that's good to go I think !

Reply:
Just to check I've really understood the logic here. The `non_mir_emits` is true if some other `--emit` has been found and you need to infer the output file. In `non_mir_emits` is false, there's only `--emit mir` and the output file does not need anything special. Am I right?

Reply:
I'm still a bit cautious about the possible race here (as seen in https://github.com/compiler-explorer/compiler-explorer/pull/3064 for example). And also, correct me if I'm wrong, but it looks like we only need to the second execution to get both the assembly and the MIR output. If so, we could simply do only one exec, saving some resources and avoiding a possible race on the assembly file. What do you think?

Reply:
I don't know how the the execution happens under the hood, so yeah there may be a possible race condition I'm not aware of. I just thought it was going to be fine because this is replicating the same behavior of llvm ir output so I thought it was already being taken care of by the "execution environment". 

And you are right, you could get both the asm and the mir in one single compiler invocation and for that matter also include the llvm ir. I didn't feel confident enough to do that sort of refactoring but I agree that would be the best way to go.

Reply:
> Just to check I've really understood the logic here. The `non_mir_emits` is true if some other `--emit` has been found and you need to infer the output file. In `non_mir_emits` is false, there's only `--emit mir` and the output file does not need anything special. Am I right?

Yeah that's exactly it.

Reply:
Ok, thank you for this, and sorry this is taking longer than it should. I'll take some more time to check what's best and maybe respin my other PR for making the single exec here easier.

Reply:
That's totally ok, I agree a single execution will be better. And I already have it running locally so I can play with it myself for my own personal benefit ;)